x-pack/plugins/features/server/feature_schema.ts

```diff
@@ -61,73 +99,190 @@ const schema = Joi.object({
     .required(),
   management: managementSchema,
   catalogue: catalogueSchema,
-  privileges: Joi.object({
-    all: privilegeSchema,
-    read: privilegeSchema,
-  }).required(),
+  privileges: Joi.alternatives(
+    Joi.string().allow('none'),
+    Joi.object({
+      all: privilegeSchema,
+      read: privilegeSchema,
+    })
+  ).required(),
+  subFeatures: Joi.when('privileges', {
+    is: 'none',
+    then: Joi.array()
+      .items(subFeatureSchema)
+      .max(0),
+    otherwise: Joi.array().items(subFeatureSchema),
+  }),
   privilegesTooltip: Joi.string(),
   reserved: Joi.object({
     privilege: privilegeSchema.required(),
     description: Joi.string().required(),
   }),
 });
 
-export function validateFeature(feature: FeatureWithAllOrReadPrivileges) {
+export function validateFeature(feature: IFeature) {
   const validateResult = Joi.validate(feature, schema);
   if (validateResult.error) {
     throw validateResult.error;
   }
   // the following validation can't be enforced by the Joi schema, since it'd require us looking "up" the object graph for the list of valid value, which they explicitly forbid.
   const { app = [], management = {}, catalogue = [] } = feature;
 
-  const privilegeEntries = [...Object.entries(feature.privileges)];
-  if (feature.reserved) {
-    privilegeEntries.push(['reserved', feature.reserved.privilege]);
-  }
+  const featureApps = new Set(app);
```

Comment:
Makes sense to me!

Reply:
__nit:__ sorry for hijacking the thread with insignificant detail, just noticed code excerpt in email notification and decided to post it here to not forget :slightly_smiling_face: 
```suggestion
  const featureApps = new Set<string>(app);
```

Otherwise it's `Set<any>`.

Reply:
@azasypkin it _appears_ to correctly infer `Set<string>`:

![image](https://user-images.githubusercontent.com/3493255/75253411-7bb9ff80-57ac-11ea-9c49-791cee22121b.png)

Reply:
Ha! You're right, I see this signature in  `lib.es2015.collection.d.ts`:
```ts
interface SetConstructor {
    new <T = any>(values?: readonly T[] | null): Set<T>;
    readonly prototype: Set<any>;
}
```
It seems I need to tune my Webstorm, sorry for the noise :see_no_evil: